### ISSUE TITLE
"ValueError: not enough values to unpack" handling JOIN message

### ISSUE THREAD

**Issue by alanbriolat** [HackSoc/csbot#148]
It looks like the asynchronous nature of processing messages/events in bot means related events can happen out-of-order. In this particular example, I think the bot expects the `extended-join` "capability" to have been enabled before the first time it receives a join message, however it joins the channels faster than it can negotiate capabilities.

```
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'AUTHENTICATE <REDACTED>'
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'CAP LS'
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'CAP END'
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'CAP REQ :sasl'
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'USER csyorkbot * * :Mathison'
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'NICK Mathison'
[2019/02/14 13:31:08] (D:csbot.raw_log) <<< 'AUTHENTICATE PLAIN'
[2019/02/14 13:31:08] (D:csbot.raw_log) >>> ':card.freenode.net NOTICE * :*** Looking up your hostname...'
[2019/02/14 13:31:09] (D:csbot.raw_log) >>> ':card.freenode.net NOTICE * :*** Checking Ident'
[2019/02/14 13:31:09] (D:csbot.raw_log) >>> ':card.freenode.net NOTICE * :*** Found your hostname'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net NOTICE * :*** No Ident response'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net CAP * ACK :sasl '
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 250 Mathison :Highest connection count: 6375 (6374 clients) (1709589 connections received)'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 251 Mathison :There are 114 users and 90100 invisible on 32 servers'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 005 Mathison CHARSET=ascii NICKLEN=16 CHANNELLEN=50 TOPICLEN=390 DEAF=D FNC TARGMAX=NAMES:1,LIST:1,KICK:1,WHOIS:1,PRIVMSG:4,NOTICE:4,ACCEPT:,MONITOR: EXTBAN=$,ajrxz CLIENTVER=3.0 KNOCK CPRIVMSG CNOTICE :are supported by this server'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 252 Mathison 36 :IRC Operators online'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 004 Mathison card.freenode.net ircd-seven-1.1.7 DOQRSZaghilopswz CFILMPQSbcefgijklmnopqrstvz bkloveqjfI'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 005 Mathison WHOX SAFELIST ELIST=CTU ETRACE :are supported by this server'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 002 Mathison :Your host is card.freenode.net[38.229.70.22/6667], running version ircd-seven-1.1.7'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 906 Mathison :SASL authentication aborted'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 253 Mathison 21 :unknown connection(s)'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 001 Mathison :Welcome to the freenode Internet Relay Chat Network Mathison'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 003 Mathison :This server was created Thu Aug 23 2018 at 19:18:42 UTC'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 266 Mathison 90214 94264 :Current global users 90214, max 94264'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 265 Mathison 4845 6374 :Current local users 4845, max 6374'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 255 Mathison :I have 4845 clients and 1 servers'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 005 Mathison CHANTYPES=# EXCEPTS INVEX CHANMODES=eIbq,k,flj,CFLMPQScgimnprstz CHANLIMIT=#:120 PREFIX=(ov)@+ MAXLIST=bqeI:100 MODES=4 NETWORK=freenode STATUSMSG=@+ CALLERID=g CASEMAPPING=rfc1459 :are supported by this server'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 254 Mathison 51813 :channels formed'
[2019/02/14 13:31:18] (D:csbot.raw_log) <<< 'JOIN #cs-york'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':Mathison MODE Mathison :+i'
[2019/02/14 13:31:18] (D:csbot.raw_log) <<< 'JOIN #compsoc-uk'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 376 Mathison :End of /MOTD command.'
[2019/02/14 13:31:18] (D:csbot.raw_log) <<< 'JOIN #hacksoc'
[2019/02/14 13:31:18] (D:csbot.raw_log) <<< 'JOIN #cs-york-dev'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 900 Mathison Mathison!~csyorkbot@unaffiliated/hjmills/bot/mathison Mathison :You are now logged in as Mathison.'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':NickServ!NickServ@services. NOTICE Mathison :This nickname is registered. Please choose a different nickname, or identify via \x02/msg NickServ identify <password>\x02.'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':NickServ!NickServ@services. NOTICE Mathison :This nickname is registered. Please choose a different nickname, or identify via \x02/msg NickServ identify <password>\x02.'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':freenode-connect!frigg@freenode/utility-bot/frigg NOTICE Mathison :Welcome to freenode. To protect the network all new connections will be scanned for vulnerabilities. This will not harm your computer, and vulnerable hosts will be notified.'
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':freenode-connect!frigg@freenode/utility-bot/frigg PRIVMSG Mathison :\x01VERSION\x01'
[2019/02/14 13:31:24] (D:csbot.raw_log) >>> ':card.freenode.net CAP Mathison LS :account-notify extended-join identify-msg multi-prefix sasl'
Traceback (most recent call last):
  File "/app/run_csbot.py", line 3, in <module>
    main(auto_envvar_prefix='CSBOT')
  File "/venv/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/venv/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/venv/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/venv/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/app/csbot/__init__.py", line 96, in main
    client.loop.run_until_complete(client.run())
  File "/usr/lib/python3.6/asyncio/base_events.py", line 473, in run_until_complete
    return future.result()
  File "/app/csbot/irc.py", line 270, in run
    await self.read_loop()
  File "/app/csbot/irc.py", line 309, in read_loop
    self.line_received(self.codec.decode(line[:-2]))
  File "/app/csbot/core.py", line 219, in line_received
    super().line_received(line)
  File "/app/csbot/irc.py", line 366, in line_received
    self.message_received(msg)
  File "/app/csbot/irc.py", line 370, in message_received
    self._dispatch_method('irc_' + msg.command_name, msg)
  File "/app/csbot/irc.py", line 721, in _dispatch_method
    method(*args, **kwargs)
  File "/app/csbot/core.py", line 372, in irc_JOIN
    channel, account, _ = msg.params
ValueError: not enough values to unpack (expected 3, got 1)
```

**Comment by LordAro** [comment]
Given `irc_JOIN` is guarded by:
https://github.com/HackSoc/csbot/blob/8503fd508def9106c301bb870df67045b3171c20/csbot/core.py#L369-L371

I must agree, this seems to be due to capabilities being requested and received after the initial join messages were sent

Also, I note this in the log:
```
[2019/02/14 13:31:18] (D:csbot.raw_log) >>> ':card.freenode.net 906 Mathison :SASL authentication aborted'
```

Is this due to sending an AUTHENTICATE before requesting the sasl CAP?

Connection setup needs to be a bit less concurrent :)

**Comment by alanbriolat** [comment]
Fixed in #155.
